proposals/0002-type-states.md

```diff
@@ -0,0 +1,227 @@
+
+We introduce the concept of ‘Type States’, by which the matching of present and expected ‘states’ of the contract can be enforced automatically by the Flint language. Runtime checks could be removed for internal calls where the state would already be known. When the state is not known (during external calls, for example) calls would have a state check behind the scenes. Overall this would effectively result in the contract becoming a state machine, enforced by the compiler and type system.
+
+By way of implementation, the contract is given a ‘state’ taking a predefined, enumerated, set of values, and function declarations give the acceptable states for calls to that function.
+Functions are then able to set the state as they like, and calls to functions with incorrect state are rejected.
+
+It would of course be up to the user to utilise the new concept, and if unsafe behaviour is desired then it would be allowed. The programmer could elect to ignore the concept of states entirely, or simply use ordinary contract properties and assertions at the start of function declarations to maintain their own state.
+
+```swift
+contract Auction (Preparing, InProgress, Terminated) { // Enumeration of states.
+  var beneficiary: Address
+  var highestBidder: Address
+  var highestBid: Wei
+}
+
+Auction @(any) :: caller <- (any) {
+  public init() {
+    self.beneficiary = caller
+    self.highestBidder = caller
+    self.highestBid = Wei(0)
+    become Preparing
+  }
+}
+
+Auction @(Preparing) :: (beneficiary) {
+  public mutating func setBeneficiary(beneficiary: Address) {
+    self.beneficiary = beneficiary
+  }
+
+  mutating func openAuction() -> Bool {
+    // ...
+    return true
+    become InProgress
+  }
+}
+Auction @(InProgress) :: (beneficiary) {
+
+  mutating func endAuction() {
+    // ...
+    become Terminated
+  }
+}
+
+Auction @(InProgress) :: (any) {
+  @payable
+  func bid(implicit value: Wei) -> Bool{
+    // ...
+    // State is not explicitly changed.
+    return Bool
+  }
+}
+```
+## Motivation
+
+Insufficient and incorrect state management in Solidity code have led to security vulnerabilities and unexpected behaviour in widely deployed smart contracts. Avoiding these vulnerabilities by the design of the language is a strong advantage.
+
+The first Parity Multi-Sig wallet hack was the result of a function being called when the contract was in the wrong state. Specifically the initWallet function was supposed to be called exclusively in the “initialization” phase of the smart contract. However, this function could actually be called at any stage in the lifetime of the contract, and allowed for an attacker to set themselves as the owner after the contract had already been initialised and Ether deposited, and therefore steal Ether from the wallet. The value of Ether lost at the time amounted to around $260 Million.
+```javascript
+// Constructor before fix was delgated this this sub-function that was public by default
```

Comment:
This comment makes no sense. I'm not sure if this would be better or not, as the hack is not simple to explain with the delegation.

> In the vulnerable code, the constructor delegated to this sub-function which was public by default.